fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,1162 @@
+				}
+				if previous.C == l || previous.C == r {
+					inc++
+				}
+			}
+			if inc < 3 {
+				o := p.orientation(l, r)
+				orient := big.NewFloat(o)
+				if orient.Cmp(Zero) < 0 {
+					// p is on the other side of the line formed by the two points
+					// therefore cross the edge
+					previous = t
+					t = nil
+					for _, t1 := range l.adjacentTriangles {
+						for _, t2 := range r.adjacentTriangles {
+							if t1.Equals(t2) && !t1.Equals(previous) {
+								t = t1
+								break
+							}
+						}
+					}
+					if t == nil {
+						// if t is nil the point is outside the current triangulation
+						return previous, false, l, r
+					}
+					found = false
+					break
+				} else if orient.Cmp(Zero) == 0 {
+					ab := big.NewFloat(p.orientation(t.A, t.B))
+					bc := big.NewFloat(p.orientation(t.B, t.C))
+					ca := big.NewFloat(p.orientation(t.C, t.A))
+					// p is on the edge if it is on the line formed by the points and if it is in between the 2 other edges
+					// in that triangle
+					if ab.Cmp(Zero) >= 0 && bc.Cmp(Zero) >= 0 && ca.Cmp(Zero) >= 0 {
+						return t, true, nil, nil
+					}
+				}
+			}
+		}
+	}
+	return t, false, nil, nil
+}
+
+// findTriangle goes down the hierarchy to find the triangle in which the point is located.
+// It returns the triangle and whether a point is on an edge.
+func findTriangle(t *Triangle, p *Point) (*Triangle, bool) {
+	// get information about the point in respect to its position to the triangle.
+	inside, edge := p.inTriangle(t)
+	if !inside {
+		return nil, false
+	}
+	// leaf triangle
+	if len(t.children) == 0 {
+		return t, edge
+	}
+	for _, tc := range t.children {
+		tt, oe := findTriangle(tc, p)
+		// if tt is nil then look at other children. Otherwise return the tt.
+		if tt != nil {
+			return tt, oe
+		}
+	}
+	return nil, false
+}
+
+// insertP inserts a point inside a triangle
+func (d *Delaunay) insertP(t *Triangle, p *Point) {
+	// form three new triangles
+	t1 := NewTriangle(t.A, t.B, p)
+	t1.isInTriangulation = true
+	t2 := NewTriangle(t.B, t.C, p)
+	t2.isInTriangulation = true
+	t3 := NewTriangle(t.A, p, t.C)
+	t3.isInTriangulation = true
+	// adjust the adjacent triangles for all points involved
+	p.adjacentTriangles = p.adjacentTriangles.append(t1, t2, t3)
+	t.isInTriangulation = false
+	t.A.adjacentTriangles = t.A.adjacentTriangles.remove(t)
+	t.B.adjacentTriangles = t.B.adjacentTriangles.remove(t)
+	t.C.adjacentTriangles = t.C.adjacentTriangles.remove(t)
+	t.A.adjacentTriangles = t.A.adjacentTriangles.append(t1, t3)
+	t.B.adjacentTriangles = t.B.adjacentTriangles.append(t1, t2)
+	t.C.adjacentTriangles = t.C.adjacentTriangles.append(t2, t3)
+	if d.useHierarchical {
+		t.children = append(t.children, t1, t2, t3)
+	}
+	d.triangles = append(d.triangles, t1, t2, t3)
+	// change the edges so it is a valid delaunay triangulation
+	d.validateEdge(t1, p)
+	d.validateEdge(t2, p)
+	d.validateEdge(t3, p)
+}
+
+func (d *Delaunay) insertPonBorderE(t *Triangle, p *Point) {
```

Comment:
This seems very similar to `insertPonE`, there needs to be documentation explaining what these do.

`func (d *Delaunay) insertAtBorderEdge(new *Point, t *Triangle) {`